src/js/legend.js

```diff
@@ -0,0 +1,215 @@
+
+    // Return a D3 color scale based on user supplied colors or the d3.scale.category10() defaults
+
+    gpii.chartAuthoring.pieChart.legend.getColorScale = function (colors) {
+      return colors ? d3.scale.ordinal().range(colors) : d3.scale.category10();
+    };
+
+    // Takes the dataSet array and the color array, and returns a consolidated object array to ease sorting and other operations while keeping colors "correct"
+
+    gpii.chartAuthoring.pieChart.legend.consolidateDataAndColors = function (model) {
+      var dataSet = model.dataSet;
+      var color = gpii.chartAuthoring.pieChart.legend.getColorScale(model.colors);
+
+        var c = [];
+        for(var i=0; i<dataSet.length; i++) {
+            var d = {
+                id: dataSet[i].id,
+                label: dataSet[i].label,
+                value: dataSet[i].value,
+                color: color(i)
+            };
+            c.push(d);
+        }
+        return c;
+
+    };
+
+    gpii.chartAuthoring.pieChart.legend.draw = function (that) {        
+        var table = that.table,
+            l = that.options.legendOptions,
+            dataSet = that.model.dataSetWithColors,
+            rowClass = that.classes.row,
+            legendColorCellClass = that.classes.legendColorCell,
+            legendLabelCellClass = that.classes.labelCell,
+            legendValueCellClass = that.classes.valueCell,
+            sort = l.sort;
+        var tbody = table.selectAll("tbody");
+
+        var rows = tbody.selectAll("tr")
+            .data(dataSet, function (d) {
+                return d.id;
+            });
+
+        var addedRows = rows.enter().append("tr");
+
+        addedRows
+        .attr({
+            "class": rowClass
+        });
+
+        addedRows
+        .append("td")
+        .attr({
+            "style": function (d) {
+                return "background-color: "+d.color+";";
+            },
+            "class": legendColorCellClass
+        });
+
+        addedRows
+        .append("td")
+        .attr({
+            "class": legendLabelCellClass
+        })
+        .html(function (d) {
+            return d.label;
+        });
+
+        addedRows.append("td")
+        .attr({
+            "class": legendValueCellClass
+        })
+        .html(function (d) {
+            return d.value;
+        });
+
+        var removedRows = rows.exit();
+
+        removedRows.remove();
+
+        if(sort === true) {
+            rows.sort(function (a,b){
+                return b.value - a.value;
+            });
+        }
+    };
+
+    gpii.chartAuthoring.pieChart.legend.create = function (that) {
+        var container = that.container,
+            dataSet = that.model.dataSet,
+            l = that.options.legendOptions,
+            colors = l.colors,
+            tableClass = that.classes.table;
+
+        if (dataSet.length === 0) {
+            return;
+        }
+
+        that.color = gpii.chartAuthoring.pieChart.legend.getColorScale(colors);
```

Comment:
This line can be made declarative into the component defaults block.
An example - https://github.com/fluid-project/infusion/blob/master/src/components/textfieldSlider/js/TextfieldSlider.js#L120-L125